src/util/system.cpp

```diff
@@ -432,20 +432,21 @@ const fs::path& ArgsManager::GetDataDir(bool net_specific) const
         path = GetDefaultDataDir();
     }
 
-    if (!fs::exists(path)) {
-        fs::create_directories(path / "wallets");
-    }
-
     if (net_specific && !BaseParams().DataDir().empty()) {
         path /= fs::PathFromString(BaseParams().DataDir());
-        if (!fs::exists(path)) {
-            fs::create_directories(path / "wallets");
-        }
     }
 
     return path;
 }
 
+void ArgsManager::EnsureDataDir(bool net_specific)
```

Comment:
Maybe introduce two functions: one for `net_specific == true` and the other for `net_specific == true`?

It will make code cleaner on the caller sites.

Reply:
Totally agree there could be two wrapper functions here if desired, `EnsureDataDir` and `EnsureNetDataDir` perhaps? It might just means more duplicated code in the end though?

Reply:
I did some digging, and I think (would be nice to get more confirmation on this) we can scrap the `net_specific == false` version entirely, we don't seem to need it. `EnsureDataDir` is called 3 times: 2 times right before reading config files - but it looks like we don't actually need to ensure a datadir there. If datadir doesn't exist, by definition no config files can be read (and that's handled gracefully). The only time we need it is where `net_specific==true`.

Intuitively, this makes sense - it can never hurt to ensure that all the necessary datadirs (both base as well as the more specific net) exist. However, there actually is a pitfall, because `GetDataDir()` [does some caching](https://github.com/bitcoin/bitcoin/blob/8126551d54ffd290ed5767248be4b3d19243787b/src/util/system.cpp#L420-L422), we're [not allowed](https://github.com/bitcoin/bitcoin/blob/8126551d54ffd290ed5767248be4b3d19243787b/src/qt/bitcoin.cpp#L598) to call `GetDatDir(true)` before we've determined the network we're on (which requires us to first read the basedir and check if there's a config file there).

For that reason, I think calling it `EnsureDataDirNet` is probably better - and aligns with `GetDataDirNet` naming.

<details>
<summary>git diff (quick patch just to prototype)</summary>

```diff
diff --git a/src/bitcoind.cpp b/src/bitcoind.cpp
index 3a36ff700..07002cc8a 100644
--- a/src/bitcoind.cpp
+++ b/src/bitcoind.cpp
@@ -153,7 +153,6 @@ static bool AppInit(NodeContext& node, int argc, char* argv[])
         if (!CheckDataDirOption()) {
             return InitError(Untranslated(strprintf("Specified data directory \"%s\" does not exist.\n", args.GetArg("-datadir", ""))));
         }
-        args.EnsureDataDir(/*net_specific=*/false);
         if (!args.ReadConfigFiles(error, true)) {
             return InitError(Untranslated(strprintf("Error reading configuration file: %s\n", error)));
         }
diff --git a/src/qt/bitcoin.cpp b/src/qt/bitcoin.cpp
index 80e607f77..59f433749 100644
--- a/src/qt/bitcoin.cpp
+++ b/src/qt/bitcoin.cpp
@@ -596,7 +596,6 @@ int GuiMain(int argc, char* argv[])
     try {
         /// 6b. Parse bitcoin.conf
         /// - Do not call gArgs.GetDataDirNet() before this step finishes
-        gArgs.EnsureDataDir(/*net_specific=*/false);
         if (!gArgs.ReadConfigFiles(error, true)) {
             InitError(strprintf(Untranslated("Error reading configuration file: %s\n"), error));
             QMessageBox::critical(nullptr, PACKAGE_NAME,
diff --git a/src/util/system.cpp b/src/util/system.cpp
index 329100008..75fd91516 100644
--- a/src/util/system.cpp
+++ b/src/util/system.cpp
@@ -439,9 +439,9 @@ const fs::path& ArgsManager::GetDataDir(bool net_specific) const
     return path;
 }
 
-void ArgsManager::EnsureDataDir(bool net_specific)
+void ArgsManager::EnsureDataDirNet()
 {
-    fs::path path = GetDataDir(net_specific);
+    fs::path path = GetDataDir(true);
     if (!fs::exists(path)) {
         fs::create_directories(path / "wallets");
     }
@@ -492,7 +492,7 @@ bool ArgsManager::IsArgSet(const std::string& strArg) const
 
 bool ArgsManager::InitSettings(std::string& error)
 {
-    EnsureDataDir(/*net_specific=*/true);
+    EnsureDataDirNet();
     if (!GetSettingsPath()) {
         return true; // Do nothing if settings file disabled.
     }
diff --git a/src/util/system.h b/src/util/system.h
index 3bcbd5c8e..23c15228e 100644
--- a/src/util/system.h
+++ b/src/util/system.h
@@ -478,10 +478,8 @@ protected:
 
     /**
      * Create data directory if it doesn't exist
-     *
-     * @param net_specific Create a network-identified data directory
      */
-    void EnsureDataDir(bool net_specific);
+    void EnsureDataDirNet();
 
 private:
     /**
```
</details>